src/test/java/spoon/reflect/ast/CloneTest.java

```diff
@@ -265,4 +269,53 @@ public void testIssue3389() {
 		// contract: clone preserves the order
 		assertEquals(c.getFields().get(0), c.clone().getTypeMembers().get(0));
 	}
+
+	@Test
+	public void testCloneKeepsImplicitModifierState() {
+		// contract: When cloning an element the implicit state is kept as well
+		CtClass<?> ctClass = Launcher.parseClass(
+				"class Foo { interface Nested { void bar(); } }"
+		);
+		CtMethod<?> method = ctClass.getNestedType("Nested").getMethod("bar");
+
+		Set<CtExtendedModifier> expectedModifiers = new HashSet<>();
+		expectedModifiers.add(new CtExtendedModifier(ModifierKind.PUBLIC, true));
+		expectedModifiers.add(new CtExtendedModifier(ModifierKind.ABSTRACT, true));
+
+		assertEquals(
+				expectedModifiers,
+				method.getExtendedModifiers()
+		);
+
+		assertEquals(
+				expectedModifiers,
+				method.clone().getExtendedModifiers()
+		);
+	}
+
+	@Test
+	public void testCloneClonesExtendedModifiers() {
+		// contract: When cloning an element the extended modifiers are cloned as well
+		CtClass<?> ctClass = Launcher.parseClass(
+				"class Foo { interface Nested { void bar(); } }"
+		);
+		CtMethod<?> method = ctClass.getNestedType("Nested").getMethod("bar");
+
+		Set<CtExtendedModifier> expectedModifiers = new HashSet<>();
+		expectedModifiers.add(new CtExtendedModifier(ModifierKind.PUBLIC, true));
+		expectedModifiers.add(new CtExtendedModifier(ModifierKind.ABSTRACT, true));
+
+		assertEquals(
+				expectedModifiers,
+				method.getExtendedModifiers()
+		);
+
+		CtMethod<?> clone = method.clone();
+		method.getExtendedModifiers().iterator().next().setImplicit(false);
+
+		assertEquals(
+				expectedModifiers,
+				clone.getExtendedModifiers()
+		);
```

Comment:
Given that implicitness may be removed from the equality check of extended modifiers, this is rather fragile. I'd suggest altering this test to check that there are the expected amount of unique objects. Something like this:

```suggestion
		Set<CtExtendedModifier> methodModifiers = method.getExtendedModifiers();
		Set<CtExtendedModifier> clonedMethodModifiers = method.clone().getExtendedModifiers();

		Set<CtExtendedModifier> distinctModifiers = Collections.newSetFromMap(new IdentityHashMap<>());
		distinctModifiers.addAll(methodModifiers);
		distinctModifiers.addAll(clonedMethodModifiers);

		assertEquals(methodModifiers, clonedMethodModifiers);
		assertEquals(distinctModifiers.size(), methodModifiers.size() + clonedMethodModifiers.size());
```

The check on distinct modifiers could arguably be extracted into a separate helper method, but I think you get the idea.